src/controller/worker.rs

```diff
@@ -131,31 +131,36 @@ impl<A: AllocationsChecker> Worker<A> {
     }
 
     // TODO: process all chunks, not only the first one
-    async fn execute_query(
-        &self,
-        query_str: String,
-        dataset: String,
-    ) -> Result<QueryResult, QueryError> {
-        let query: BatchRequest = serde_json::from_str(query_str.as_str())
+    async fn execute_query(&self, query_str: String, dataset: String) -> QueryResult {
+        let query = sqn_query::Query::from_json_bytes(query_str.as_bytes())
             .map_err(|e| QueryError::BadRequest(format!("Couldn't parse query: {e:?}")))?;
+        let Some(first_block) = query.first_block() else {
+            return Err(QueryError::BadRequest(
+                "Query without first_block".to_owned(),
+            ));
+        };
         let chunks_guard = self
             .state_manager
-            .find_chunks(&dataset, (query.from_block as u32).into())?;
-        let path = chunks_guard.iter().next().cloned();
-        if let Some(path) = path {
-            tokio::spawn(async move {
-                let ctx = query::context::prepare_query_context(&path).await.unwrap();
-                let result = query::processor::process_query(&ctx, query).await?;
-                Ok(QueryResult::new(result, 1)?)
-            })
-            .await
-            .unwrap_or_else(|e| {
-                Err(QueryError::Other(
-                    anyhow::Error::new(e).context("Query processing task panicked"),
-                ))
+            .find_chunks(&dataset, (first_block as u32).into())?;
+        let Some(path) = chunks_guard.iter().next().cloned() else {
+            return Err(QueryError::NotFound);
+        };
+        tokio::task::spawn_blocking(move || {
+            polars_core::POOL.install(move || {
```

Comment:
Another option would be to install a global worker once, which would read from channel and fork to query execution. But then two things would probably become more complicated:
- replacing `for_each_concurrent` with a synchronous analogue,
- handling panics in individual query tasks.